src/App.tsx

```diff
@@ -1,26 +1,247 @@
-import React from 'react';
-import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import React, {
+  useEffect,
+  useRef,
+  useState,
+  useCallback,
+  useMemo,
+} from 'react';
+import * as api from './api/todos';
+import { Todo } from './types/Todo';
+import { TypeFilter } from './types/TypeFilter';
+import { TodoHeader } from './components/TodoHeader';
+import { TodoList } from './components/TodoList';
+import { TodoFooter } from './components/TodoFooter';
+import { ErrorNotifications } from './components/ErrorNotifications';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [newTodoInput, setNewTodoInput] = useState('');
+  const [error, setError] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingIds, setLoadingIds] = useState<number[]>([]);
+  const [filter, setFilter] = useState<TypeFilter>(TypeFilter.All);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const activeTodos = useMemo(
+    () => todos.filter(todo => !todo.completed),
+    [todos],
+  );
+
+  const completedTodos = useMemo(
+    () => todos.filter(todo => todo.completed),
+    [todos],
+  );
+
+  const activeCount = activeTodos.length;
+
+  const showError = useCallback((message: string) => {
+    setError(message);
+    setTimeout(() => setError(''), 3000);
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    switch (filter) {
+      case TypeFilter.Active:
+        return activeTodos;
+
+      case TypeFilter.Completed:
+        return completedTodos;
+
+      default:
+        return todos;
+    }
+  }, [filter, activeTodos, completedTodos, todos]);
+
+  useEffect(() => {
+    if (!isLoading) {
+      inputRef.current?.focus();
+    }
+  }, [isLoading]);
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      try {
+        const todosData = await api.getTodos();
+
+        setTodos(todosData);
+      } catch (err) {
+        showError('Unable to load todos');
+      }
+    };
+
+    fetchTodos();
+  }, [showError]);
+
+  // Add todo
+  const addTodo = useCallback(
+    async (e: React.FormEvent<HTMLFormElement>) => {
+      e.preventDefault();
+
+      const title = newTodoInput.trim();
+
+      if (!title) {
+        showError('Title should not be empty');
+
+        return;
+      }
+
+      setIsLoading(true);
+      const userId = api.USER_ID;
+      const todoData = { userId, title, completed: false };
+
+      setTempTodo({ id: 0, ...todoData });
+
+      try {
+        const newTodo = await api.createTodo(todoData);
+
+        setTodos(prev => [...prev, newTodo]);
+        setNewTodoInput('');
+        setTimeout(() => inputRef.current?.focus(), 0);
+      } catch (err) {
+        showError('Unable to add a todo');
+      } finally {
+        setTempTodo(null);
+        setIsLoading(false);
```

Comment:
isLoading state looks redundant.
Instead isLoading you can use tempTodo state (tempTodo and isLoading changing in same time and tempTodo technically always truthy or false as isLoding)